File: aoc-android-frontend/src/api/aoc-api-service/model/AoCReport.js
import Base from "./Base";

class AoCReportSummary extends Base {
    constructor(data) {
        super(data);
        this.id = data.id;
        this.line = data.line;
        this.snippet = data.snippet;
        this.class = data.class;
        this.aoc = data.aoc;
        this.path = data.path;
    }
}

class AoCReportDetail extends AoCReportSummary {
    constructor(data) {
        super(data);
        this.projectId = data.projectId;
        this.project = data.project;
    }
}

class AoCStatistics extends Base {
    constructor(data) {
        super(data);
        /**
         * @type {number}
         */
        this.count = data.count;
        /**
        * @type {Map<string, number>}
        */
        this.class = data.class;
        /**
         * @type {Map<string, number>}
         */
        this.aoc = data.aoc;
    }
}

class ProjectAoCStatistics extends AoCStatistics {
    constructor(data) {
        super(data);
        /**
         * @type {Map<string, AoCStatistics>}
         */
        this.project = data.project
    }
}

export {
    AoCReportSummary, AoCReportDetail, AoCStatistics, ProjectAoCStatistics
}

File: backend/model/aoc_report.py
from backend.model.base import Base

from marshmallow import Schema, fields

from sqlalchemy import Integer, String, ForeignKey
from sqlalchemy.orm import Mapped, mapped_column, relationship


class AoCReport(Base):
    __tablename__ = "aoc_reports"
    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    project_id: Mapped[int] = mapped_column(ForeignKey("projects.id"))
    project: Mapped["Project"] = relationship(
        back_populates="aoc_reports", lazy="subquery"
    )
    line: Mapped[int] = mapped_column(Integer())
    snippet: Mapped[str] = mapped_column(String(), nullable=False)
    class_: Mapped[str] = mapped_column(String(), nullable=False, name="class")
    aoc: Mapped[str] = mapped_column(String(), nullable=False)
    path: Mapped[str] = mapped_column(String(), nullable=True)


class AOSPAoCReport(Base):
    __tablename__ = "aosp_aoc_reports"
    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    project_id: Mapped[int] = mapped_column(ForeignKey("aosp_projects.id"))
    project: Mapped["AOSPProject"] = relationship(
        back_populates="aoc_reports", lazy="subquery"
    )
    line: Mapped[int] = mapped_column(Integer())
    snippet: Mapped[str] = mapped_column(String(), nullable=False)
    class_: Mapped[str] = mapped_column(String(), nullable=False, name="class")
    aoc: Mapped[str] = mapped_column(String(), nullable=False)
    path: Mapped[str] = mapped_column(String(), nullable=True)
    loc: Mapped[int] = mapped_column(Integer(), nullable=True)
    commit: Mapped[str] = mapped_column(String(), nullable=False)
